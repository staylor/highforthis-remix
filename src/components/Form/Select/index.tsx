import cn from 'classnames';
import type { ReactNode } from 'react';
import React from 'react';

import { inputBase } from '@/components/Form/styles';

type Choice = string | number | { label: string; value: string | number };
type Group = { label: string; choices: Choice[] };

type Props = {
  name?: string;
  className?: string | null;
  multiple?: boolean | string | number;
  value?: any;
  onChange?: any;
  placeholder?: string;
  choices?: Choice[];
  groups?: Group[];
  children?: ReactNode;
};

const renderOption = (choice: any) => {
  if (typeof choice === 'object') {
    return (
      <option key={choice.value} value={choice.value}>
        {choice.label}
      </option>
    );
  }
  return (
    <option key={choice} value={choice}>
      {choice}
    </option>
  );
};

export default function Select({
  className,
  placeholder,
  multiple: multipleProp = false,
  choices,
  groups,
  children = null,
  value = null,
  onChange: onChangeProp,
  ...rest
}: Props) {
  const onChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    let val: any = e.target.value;
    const multiple = Boolean(multipleProp);
    if (multiple) {
      val = [...e.target.options].filter((o) => o.selected).map((o) => o.value);
    }
    if (onChangeProp) {
      onChangeProp(val, e);
    }
  };

  return (
    <select
      {...rest}
      onChange={onChange}
      defaultValue={value}
      multiple={multipleProp ? Boolean(multipleProp) : false}
      className={cn(inputBase, className)}
    >
      {placeholder && (
        <option key={placeholder} value="">
          {placeholder}
        </option>
      )}
      {groups &&
        groups.map((group) => (
          <optgroup key={group.label} label={group.label}>
            {group.choices.map(renderOption)}
          </optgroup>
        ))}
      {choices && choices.map(renderOption)}
      {children}
    </select>
  );
}
