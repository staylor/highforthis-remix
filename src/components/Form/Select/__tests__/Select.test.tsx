import { describe, expect, test, vi } from 'vitest';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Select from '@/components/Form/Select';

const placeholder = '-- Select Flavor --';
const flavors = ['Mild', 'Hot', 'Fire'];
const flavorMap = [
  { label: 'Mild', value: 'mild' },
  { label: 'Hot', value: 'hot' },
  { label: 'Fire', value: 'fire' },
];
const tacoMap = [
  { label: 'Crunchy', value: 'crunchy' },
  { label: 'Soft', value: 'soft' },
];
const groups = [
  { label: 'Tacos', choices: tacoMap },
  { label: 'Saunce', choices: flavorMap },
];

describe('Select', () => {
  test('empty', () => {
    const { container } = render(<Select />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('add className', () => {
    const { container } = render(<Select className="foo" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('placeholder', () => {
    test('no choices', () => {
      const { container } = render(<Select placeholder={placeholder} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('choices', () => {
      const { container } = render(<Select placeholder={placeholder} choices={flavors} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('choices as objects', () => {
      const { container } = render(<Select placeholder={placeholder} choices={flavorMap} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('multiple', () => {
    test('true', () => {
      const { container } = render(<Select multiple />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('false', () => {
      const { container } = render(<Select multiple={false} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('string', () => {
      const { container } = render(<Select multiple="multiple" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('truthy', () => {
      const { container } = render(<Select multiple="1" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('falsey', () => {
      const { container } = render(<Select multiple="0" />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('onChange', () => {
    test('value', async () => {
      const user = userEvent.setup();
      const func = vi.fn();
      render(<Select onChange={func} choices={flavorMap} />);
      const value = 'fire';

      await user.selectOptions(screen.getByRole('combobox'), [value]);

      expect(func).toHaveBeenCalledWith(value, expect.anything());
    });

    test('values', async () => {
      const user = userEvent.setup();
      const selectSpy = {
        onChange: () => null,
      };

      const func = vi.spyOn(selectSpy, 'onChange');
      render(<Select multiple onChange={func as any} choices={flavorMap} />);

      await user.selectOptions(screen.getByRole('listbox'), ['mild', 'fire']);

      expect(func).toHaveBeenCalledWith(['mild', 'fire'], expect.anything());
    });
  });

  describe('choices', () => {
    test('choices', () => {
      const { container } = render(<Select choices={flavors} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('choices as objects', () => {
      const { container } = render(<Select choices={flavorMap} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('choice', () => {
      const value = 'medium';
      const { container } = render(<Select choices={flavorMap} value={value} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('groups', () => {
    test('no value', () => {
      const { container } = render(<Select groups={groups} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('value', () => {
      const value = 'mild';
      const { container } = render(<Select groups={groups} value={value} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('values', () => {
      const value = ['soft', 'mild'];
      const { container } = render(<Select multiple groups={groups} value={value} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('child nodes', () => {
    test('children', () => {
      const { container } = render(
        <Select>
          <option value="meximelt">Mexi-melt</option>
        </Select>
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    test('children with choices', () => {
      const { container } = render(
        <Select choices={flavorMap}>
          <option value="meximelt">Mexi-melt</option>
        </Select>
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
