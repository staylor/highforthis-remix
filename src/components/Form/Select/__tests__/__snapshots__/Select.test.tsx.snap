// Vitest Snapshot v1

exports[`Select > add className 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors foo"
/>
`;

exports[`Select > child nodes > children 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="meximelt"
  >
    Mexi-melt
  </option>
</select>
`;

exports[`Select > child nodes > children with choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
  <option
    value="meximelt"
  >
    Mexi-melt
  </option>
</select>
`;

exports[`Select > choices > choice 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
</select>
`;

exports[`Select > choices > choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="Mild"
  >
    Mild
  </option>
  <option
    value="Hot"
  >
    Hot
  </option>
  <option
    value="Fire"
  >
    Fire
  </option>
</select>
`;

exports[`Select > choices > choices as objects 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
</select>
`;

exports[`Select > empty 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
/>
`;

exports[`Select > groups > no value 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <optgroup
    label="Tacos"
  >
    <option
      value="crunchy"
    >
      Crunchy
    </option>
    <option
      value="soft"
    >
      Soft
    </option>
  </optgroup>
  <optgroup
    label="Saunce"
  >
    <option
      value="mild"
    >
      Mild
    </option>
    <option
      value="hot"
    >
      Hot
    </option>
    <option
      value="fire"
    >
      Fire
    </option>
  </optgroup>
</select>
`;

exports[`Select > groups > value 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <optgroup
    label="Tacos"
  >
    <option
      value="crunchy"
    >
      Crunchy
    </option>
    <option
      value="soft"
    >
      Soft
    </option>
  </optgroup>
  <optgroup
    label="Saunce"
  >
    <option
      value="mild"
    >
      Mild
    </option>
    <option
      value="hot"
    >
      Hot
    </option>
    <option
      value="fire"
    >
      Fire
    </option>
  </optgroup>
</select>
`;

exports[`Select > groups > values 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
>
  <optgroup
    label="Tacos"
  >
    <option
      value="crunchy"
    >
      Crunchy
    </option>
    <option
      value="soft"
    >
      Soft
    </option>
  </optgroup>
  <optgroup
    label="Saunce"
  >
    <option
      value="mild"
    >
      Mild
    </option>
    <option
      value="hot"
    >
      Hot
    </option>
    <option
      value="fire"
    >
      Fire
    </option>
  </optgroup>
</select>
`;

exports[`Select > multiple > false 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
/>
`;

exports[`Select > multiple > falsey 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`Select > multiple > string 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`Select > multiple > true 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`Select > multiple > truthy 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
  multiple=""
/>
`;

exports[`Select > placeholder > choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value=""
  >
    -- Select Flavor --
  </option>
  <option
    value="Mild"
  >
    Mild
  </option>
  <option
    value="Hot"
  >
    Hot
  </option>
  <option
    value="Fire"
  >
    Fire
  </option>
</select>
`;

exports[`Select > placeholder > choices as objects 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value=""
  >
    -- Select Flavor --
  </option>
  <option
    value="mild"
  >
    Mild
  </option>
  <option
    value="hot"
  >
    Hot
  </option>
  <option
    value="fire"
  >
    Fire
  </option>
</select>
`;

exports[`Select > placeholder > no choices 1`] = `
<select
  class="my-1 mr-1 last:mr-0 rounded-md border-gray-300 shadow-sm focus:border-pink focus:ring focus:ring-pink-200 focus:ring-opacity-50 transition-colors"
>
  <option
    value=""
  >
    -- Select Flavor --
  </option>
</select>
`;
